src/lib.rs

```diff
@@ -1,808 +1,268 @@
+//!     let vms = aws.connect_all()?;
+//!     let my_vm = vms.get("my_vm").unwrap();
+//!     let ssh = my_vm.ssh.as_ref().unwrap();
+//!     ssh.cmd("hostname").map(|(stdout, _)| println!("{}", stdout))?;
 //!     Ok(())
-//! }).unwrap();
+//! }
 //! ```
 //!
 //! # Live-coding
 //!
-//! The crate is under development as part of a live-coding stream series intended for users who
+//! An earlier version of this crate was written as part of a live-coding stream series intended for users who
 //! are already somewhat familiar with Rust, and who want to see something larger and more involved
 //! be built. You can find the recordings of past sessions [on
 //! YouTube](https://www.youtube.com/playlist?list=PLqbS7AVVErFgY2faCIYjJZv_RluGkTlKt).
-#![deny(missing_docs)]
+#![warn(unreachable_pub)]
+#![warn(missing_copy_implementations)]
+#![warn(trivial_casts)]
+#![warn(trivial_numeric_casts)]
+#![warn(unused_extern_crates)]
+#![warn(rust_2018_idioms)]
+#![warn(missing_debug_implementations)]
+#![allow(clippy::type_complexity)]
 
 #[macro_use]
 extern crate failure;
-extern crate rand;
-extern crate rayon;
-extern crate rusoto_core;
-extern crate rusoto_ec2;
-#[macro_use]
-extern crate scopeguard;
 #[macro_use]
 extern crate slog;
-extern crate slog_term;
-extern crate ssh2;
-extern crate tempfile;
 
-use failure::{Error, ResultExt};
-use rayon::prelude::*;
-use rusoto_core::request::HttpClient;
-use rusoto_core::Region;
-use rusoto_core::{DefaultCredentialsProvider, ProvideAwsCredentials};
+use educe::Educe;
+use failure::Error;
+use itertools::Itertools;
 use std::collections::HashMap;
-use std::io::Write;
-use std::{thread, time};
+use std::time;
 
-mod ssh;
+pub use sessh as ssh;
 pub use ssh::Session;
 
-/// A handle to an instance currently running as part of a tsunami.
-pub struct Machine {
-    /// An established SSH session to this host.
-    pub ssh: Option<ssh::Session>,
+pub mod providers;
+use providers::{Launcher, MachineSetup};
 
-    /// AWS EC2 instance type hosting this machine.
-    ///
-    /// See https://aws.amazon.com/ec2/instance-types/ for details.
-    pub instance_type: String,
-
-    /// The private IP address of this host on its designated VPC.
-    pub private_ip: String,
-
-    /// The publicly accessible hostname of this host.
+/// A handle to an instance currently running as part of a tsunami.
+///
+/// Run commands on the machine using the [`ssh::Session`] via the `ssh` field.
+#[derive(Educe)]
+#[educe(Debug)]
+pub struct Machine<'tsunami> {
+    pub nickname: String,
     pub public_dns: String,
-
-    /// The publicly accessible IP address of this host.
     pub public_ip: String,
-}
 
-/// A template for a particular machine setup in a tsunami.
-pub struct MachineSetup {
-    instance_type: String,
-    ami: String,
-    username: String,
-    setup: Box<Fn(&mut ssh::Session) -> Result<(), Error> + Sync>,
-}
+    /// If `Some(_)`, an established SSH session to this host.
+    #[educe(Debug(ignore))]
```

Comment:
Wouldn't it be better to just `impl Debug for ssh::Session`?

Reply:
That crate would then run into the same problem: `ssh2::Session` does not `impl Debug`.

Reply:
That's fine -- just `impl Debug` for it and don't include the session field :p